crates/conjure_core/src/rules/matrix/bubble.rs

```diff
@@ -0,0 +1,85 @@
+use conjure_core::ast::Expression;
+use conjure_core::metadata::Metadata;
+use conjure_core::rule_engine::{
+    register_rule, ApplicationError, ApplicationError::RuleNotApplicable, ApplicationResult,
+    Reduction,
+};
+use itertools::{izip, Itertools as _};
+
+use crate::ast::{Domain, SymbolTable};
+use crate::{bug, into_matrix_expr};
+
+/// Converts an unsafe index to a safe index using a bubble expression.
+#[register_rule(("Bubble", 6000))]
+fn index_to_bubble(expr: &Expression, symbols: &SymbolTable) -> ApplicationResult {
+    let Expression::UnsafeIndex(_, subject, indices) = expr else {
+        return Err(RuleNotApplicable);
+    };
+
+    let domain = subject
+        .domain_of(symbols)
+        .ok_or(ApplicationError::DomainError)?;
+
+    let Domain::DomainMatrix(_, index_domains) = domain else {
+        bug!("subject of an index expression should have a matrix domain. subject: {:?}, with domain: {:?}",subject,domain);
+    };
+
+    assert_eq!(index_domains.len(),indices.len(),"in an index expression, there should be the same number of indices as the subject has index domains");
+
+    let bubble_constraints = Box::new(into_matrix_expr![izip!(index_domains, indices)
+        .map(|(domain, index)| {
+            Expression::InDomain(Metadata::new(), Box::new(index.clone()), domain)
+        })
+        .collect_vec()]);
+
+    let new_expr = Box::new(Expression::SafeIndex(
+        Metadata::new(),
+        subject.clone(),
+        indices.clone(),
+    ));
+
+    Ok(Reduction::pure(Expression::Bubble(
+        Metadata::new(),
+        new_expr,
+        Box::new(Expression::And(Metadata::new(), bubble_constraints)),
+    )))
+}
+
+/// Converts an unsafe slice to a safe slice using a bubble expression.
+#[register_rule(("Bubble", 6000))]
+fn slice_to_bubble(expr: &Expression, symbols: &SymbolTable) -> ApplicationResult {
+    let Expression::UnsafeSlice(_, subject, indices) = expr else {
+        return Err(RuleNotApplicable);
+    };
+
+    let domain = subject
+        .domain_of(symbols)
+        .ok_or(ApplicationError::DomainError)?;
+
+    let Domain::DomainMatrix(_, index_domains) = domain else {
+        bug!("subject of a slice expression should have a matrix domain. subject: {:?}, with domain: {:?}",subject,domain);
```

Comment:
I am surprised this passes hygiene (formatting) due to whitespace.

Reply:
Do you mean the commas? I think thats because its inside a macro.

Fixed.

Reply:
I didn't realise macros were not formatted, good to know!